### PR TITLE
removing report_ration calls

### LINKED ISSUE #1436: Runtime profiling summary as of 20-Mar-2024 (using main branch and default metadata)
## Urgency Level

- [ ] Urgent: a critical feature or fix is needed
- [ ] High: the issue significantly impacts the user experience
- [ ] Medium: the issue affects functionality but has a workaround
- [ ] Low: the issue doesn't have a significant immediate impact
- [ ] Wishful: the issue should not be prioritized over any others
- [ ] Unsure: the urgency level has not been or cannot be assessed at this time 

## Status

I created a separate snapshot branch to capture the current state of the codebase — `mar-20-2024-snapshot`.

To generate a simple runtime profile, I follow these commands

```
python -m cProfile -o rufas.prof main.py

# Install snakeviz, if necessary, to visualize the results
python -m pip install snakeviz

snakeviz rufas.prof
```

image

image

At first glance, the following places are worth reviewing.

- Ration driver

image

- Cow class: `milking_update()`, `execute_ed_tai_protocol`, `preg_update()`

image

- The way the manure module reports variables may involve too many separate calls to `add_variable()` that need `deepcopy()`. This would happen when reporting non-primitive values like dictionaries.
 
image

To look at things more numerically,

(There are many more, the ones listed here are ordered by `tottime` in descending order)
image

1. `ncalls`: The number of calls made to the specific function. If there are two numbers separated by a slash (/), it means the function was called recursively. The first number represents the total number of calls, and the second number represents the number of primitive (non-recursive) calls.

2. `tottime`: The total time spent in the given function, **excluding** the time spent in sub-functions that were called by this function.

3. `percall`: The average time spent per call for the given function. It is calculated as `tottime` divided by the number of primitive calls (`ncalls`).

4. `cumtime`: The cumulative time spent in the given function, including the time spent in sub-functions that were called by this function. Should be greater than or equal to `tottime`.

5. `percall`: The average cumulative time spent per call for the given function. It is calculated as `cumtime` divided by the number of primitive calls (`ncalls`).

6. `filename:lineno(function)`: The name of the file, the line number, and the function name where the profiling information was recorded.

For example,

- The function `ration_driver.py:388(report_ration)` was called 461,674 times. It took a total of 14.14 seconds (excluding sub-function time) and a cumulative time of 14.21 seconds (including sub-function time). The average time per call was approximately 3.063e-05 seconds.

- The function `output_manager.py:130(_add_to_pool)` was called 2,782,202 times. It took a total of 4.303 seconds (excluding sub-function time) and a cumulative time of 10.23 seconds (including sub-function time). The average time per call was approximately 1.547e-06 seconds.

- The function `copy.py:128(deepcopy)` was called 7,879,274 times in total, with 4,203,224 primitive (non-recursive) calls. It took a total of 4.12 seconds (excluding sub-function time) and a cumulative time of 7.225 seconds (including sub-function time). The average time per primitive call was approximately 9.802e-06 seconds.

## Intention

To get a better understanding of where the bottlenecks are in the current codebase and potentially find some valid fixes.

## Request

- [ ] Review `ration_driver.py`.
- [ ] Review some methods in `cow.py`: `milking_update()`, `execute_ed_tai_protocol`, `preg_update()`
- [ ] Review the way the manure module reports variables to the OM.

### PR DESCRIPTION
**Pull request by JoeWaddell**
### Context

Issue(s) closed by this pull request: Won't close #1436, but should help.

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.
- [x] `default_animal.json` has been reverted to default

### What

The number of calls to RationReporter.report_ration is reduced in this PR. Instead, we're accessing Pen attributes that store the information calculated by report_ration.

### Why

RationReporter.report_ration shouldn't be called so many times. The method itself can of course be improved (it's old by RuFaS standards, so it's not terribly pythonic) to speed things up, but the information from that call is saved as pen attributes (pen.ration_nutrient_amount and pen.ration_nutrient_conc), so it shouldn't need to be called again in calf_manure_excretion/dry_cow_manure_excretion/growing_cow_manure_excretion/lactating_cow_manure_excretion.

### How

In addition to removing calls to RationReporter.report_ration in various manure calculations, also added type hints and updated docstrings as appropriate, to reduce Mypy errors. Also updated unit tests accordingly.

### Test plan

This change seems to have cut the time to run a 10X default herd simulation (e.g. all default metadata settings, except multiplying the number of animals by 10) from 952s to 498s. I haven't noted any differences in the manure outputs, but more testing may be needed. 

### 

### Other Issues
This will complete much of #1218 and #1219 , as it removes the need for a `Feed` object and `ration_formulation`.

### Ongoing problems to try to fix during review
Challenging Mypy errors: 
1) In the file Pen.py, in `_calc_animal_manure_excretion`, it gives a complex error for the call to `calc_manure_excretion`, because there is a separate method for each animal class, and for lactating cows there are more input args required than for other classes. This flags a Mypy error where there are too many calls for Calf/HeiferI-III objects, but too few for cows. This logic is handled by `_calc_animal_manure_excretion`, where the animal object is routed to the correct call setup, but Mypy flags it anyway.
Question: Should this method be A) refactored in this PR, B) should I open an Issue, or C) this Mypy error should be flagged to be ignored?

2) The method `calc_manure_excretion` in Cow.py gives the following error: Signature of "calc_manure_excretion" incompatible with supertype "HeiferIII"Mypy[override](https://mypy.readthedocs.io/en/latest/_refs.html#code-override)
Question: Should this error be officially ignored (e.g. flagging it in the code) while any changes to Animal objects (specifically related to inheritance) are made, or should we simply informally ignore it (e.g. allow the error, and open an Issue)?

### CODE REVIEW

File: RUFAS/routines/animal/life_cycle/calf.py [hunk 1]
```diff
@@ -138,16 +138,20 @@ def calc_nutrient_rqmts(self, feed: Dict[str, float], temp: float) -> None:
...
 
         Parameters
         ----------
-        feed: Dict[str, float]
-            instance of the Feed class
         methane_model : str
-            methane model used for methane emission calculations
+            Methane model used for methane emission calculations, including Boadi, IPCC.
+        nutrient_amount : Dict[str, float]
+            Amount of nutrients in pen ration, calculated per animal.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/calf.py]: Is there a unit for this?

File: RUFAS/routines/animal/life_cycle/calf.py [hunk 2]
```diff
@@ -138,16 +138,20 @@ def calc_nutrient_rqmts(self, feed: Dict[str, float], temp: float) -> None:
...
+        nutrient_conc : Dict[str, float]
+            Concentration of nutrients in pen ration, calculated per animal.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/calf.py]: Unit?

File: RUFAS/routines/animal/life_cycle/cow.py [hunk 3]
```diff
@@ -452,25 +452,34 @@ def get_user_defined_milk_protein_percent(self) -> float:
...
-            methane_model: methane model used for methane emission calculations
-            ME_intake: metabolizable energy intake, Mcal/kg DM
+        Parameters
+        ----------
+        methane_model : str
+            Methane model used for methane emission calculations, including Boadi, IPCC.
+        methane_mitigation_method: str
+            Methane mitigation method used.
+        methane_mitigation_additive_amount: float
+            Amount of methane mitigation additive, mg/kg dry matter intake (DMI).
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/cow.py]:
```suggestion
            Amount of methane mitigation additive per kg dry matter intake (DMI) (mg/kg).
```

File: RUFAS/routines/animal/life_cycle/cow.py [hunk 4]
```diff
@@ -452,25 +452,34 @@ def get_user_defined_milk_protein_percent(self) -> float:
...
+        ME_intake : float
+            Metabolizable energy intake, Mcal/kg DM
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/cow.py]:
```suggestion
            Metabolizable energy intake per kg DMI (Mcal/kg).
```

File: RUFAS/routines/animal/life_cycle/cow.py [hunk 5]
```diff
@@ -452,25 +452,34 @@ def get_user_defined_milk_protein_percent(self) -> float:
...
+        nutrient_amount : Dict[str, float]
+            Amount of nutrients in pen ration, calculated per animal.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/cow.py]: Unit?

File: RUFAS/routines/animal/life_cycle/cow.py [hunk 6]
```diff
@@ -452,25 +452,34 @@ def get_user_defined_milk_protein_percent(self) -> float:
...
+        nutrient_conc : Dict[str, float]
+            Concentration of nutrients in pen ration, calculated per animal.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/cow.py]: Unit?

File: RUFAS/routines/animal/life_cycle/heiferI.py [hunk 7]
```diff
@@ -84,30 +85,39 @@ def set_nutrient_rqmts(
...
 
-        Args:
-                feed: instance of the Feed class
-                methane_model: methane model used for methane emission calculations
+        Parameters
+        ----------
+        methane_model : str
+            Methane model used for methane emission calculations, including Boadi, IPCC.
+        nutrient_amount : Dict[str, float]
+            Amount of nutrients in pen ration, calculated per animal.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/heiferI.py]: Unit?

File: RUFAS/routines/animal/life_cycle/heiferI.py [hunk 8]
```diff
@@ -84,30 +85,39 @@ def set_nutrient_rqmts(
...
+        nutrient_conc : Dict[str, float]
+            Concentration of nutrients in pen ration, calculated per animal.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/heiferI.py]: Unit?

File: RUFAS/routines/animal/life_cycle/heiferI.py [hunk 9]
```diff
@@ -84,30 +85,39 @@ def set_nutrient_rqmts(
...
+    def phosphorus_rqmts(self, DMI: float) -> None:
         """
         Calculates and sets the animal's phosphorus requirement.
 
-        Args:
-            DMI: the Dry Matter Intake (kg)
+        Parameters
+        ----------
+        DMI : float
+            The Dry Matter Intake (kg)
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/heiferI.py]:
```suggestion
            The Dry Matter Intake (kg).
```

File: RUFAS/routines/animal/life_cycle/heiferI.py [hunk 10]
```diff
@@ -154,18 +164,23 @@ def get_non_preg_bw_change(self):
...
+        ----------
+        sim_day : int
+            Day of simulation.
 
         Returns
         -------
-
-        the second stage of heifer -- breeding stage starts
+        bool
+            The second stage of heifer -- breeding stage starts.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/heiferI.py]: I like what you had for the same boolean for heiferIIIs moving to cows. Something along the lines of:

```suggestion
            True if should be moved to second stage of heifer -- breeding stage start.
```

**Review by ew3361zh — APPROVED** [review]
LGTM! Time on main running the default scenario was reduced by almost 25% running the same scenario on this branch!

This ran successfully on all other metadatas.

There were a number of updates to the docstrings and I added some comments/suggestions there but nothing vital that would prevent this being merged. Nice work!

File: RUFAS/routines/animal/animal_module_constants.py [hunk 11]
```diff
@@ -3,95 +3,95 @@ class AnimalModuleConstants:
...
+    MINIMUM_CALCIUM: float = 0.0
     """Minimum calcium estimate, g/day"""
 
-    MILK_REDUCTION_KG = 0.25
+    MILK_REDUCTION_KG: float = 0.25
     """Milk reduction amount for each failed ration optimization attempt, kg"""
 
-    MINIMUM_HEIFER_BW_CHANGE = 0.5
+    MINIMUM_HEIFER_BW_CHANGE: float = 0.5
     """Minimum daily growth for heifers, kg."""
```
Comment (tomhuhh) [inline RUFAS/routines/animal/animal_module_constants.py:97]: Great efforts on updating the doc string for animal module constants! 

Minor suggestion: Consider adding 'daily' into the variable name? For example, 'MINIMUM_HEIFER_DAILY_GROWTH_RATE'. Also the unit should be kg/day to more precisely reflect the meaning of the variable.

File: RUFAS/routines/animal/animal_module_constants.py [hunk 12]
```diff
@@ -3,95 +3,95 @@ class AnimalModuleConstants:
...
+    DMI_CONSTRAINT_PERCENT: float = 0.20
     """The +/- percentage of DMI estimated allowed for ration formulation"""
 
-    MINIMUM_DMI = 1.0
+    MINIMUM_DMI: float = 1.0
     """Minimum estimated DMI instituted for all animals, kg/day"""
 
-    MINIMUM_DMI_PERCENTAGE = 0.01
+    MINIMUM_DMI_PERCENTAGE: float = 0.01
     """Minimum estimated DMI (kg/day), as a percentage of body_weight in kg"""
```
Comment (tomhuhh) [inline RUFAS/routines/animal/animal_module_constants.py:85]: Minor suggestion/concern: Consider the name 'MINIMUM_DAILY_DMI_RATIO' instead? It is not a percentage (0.01%), instead it is a multiplier (1%) of the body weight (kg) to calculate estimate for intake in the animal_requirements.py file under the calculate_NRC_DMI and calculate_NASEM_DMI methods.

File: RUFAS/routines/animal/life_cycle/heiferI.py [hunk 13]
```diff
@@ -154,18 +164,23 @@ def get_non_preg_bw_change(self):
             AnimalModuleConstants.MINIMUM_HEIFER_BW_CHANGE,
         )
 
-    def update(self, sim_day):
+    def update(self, sim_day: int) -> bool:
         """
         Controls heifer's grow with average daily gain based on non-preg ADG based on NRC.
         Here is the place to change growth rate with heifer feeding methods later
-        when we have heifer nutrition from the ration furmulation module.
+        when we have heifer nutrition from the ration formulation module.
```
Comment (tomhuhh) [inline RUFAS/routines/animal/life_cycle/heiferI.py:185]: Nice catch!

**Review by tomhuhh — CHANGES_REQUESTED** [review]
Great work Joe! Sorry for the delay in my review. 

I think this is a great use of existing resources in the code to make the simulation process more efficient. I'm also amazed by your efforts towards bringing the code documentation up to standard. 

I tested with manure management scenario 6 to check the manure collected by the manure handler and reception pit to see if it is in alignment with a run (with the same animal input file) that I made early today on a different branch so to verify if the manure output still remains the same after the updates proposed. The results look promising. 

On an irrelevant topic, I made some minor suggestions to the doc string updates that are included in this PR. I noticed that Niko has made some nice suggestions too in terms of the doc string. I believe we can easily address them here. 

Lastly, I would appreciate some SWE's expertise on the questions/concerns Joe brought up in the 'Ongoing problems' section of the PR description. 

Thanks!

**Comment by tomhuhh** [comment]
Sharing some results from my testing. They look identical. 

This is from earlier: 
image

image

This is from this PR: 
default_reception-pit-total-degradable-solids-04-Apr-2024_Thu_11-46-00

default_reception-pit-total-non-degradable-solids-04-Apr-2024_Thu_11-46-00

File: RUFAS/routines/animal/life_cycle/calf.py [hunk 14]
```diff
@@ -138,31 +138,49 @@ def calc_nutrient_rqmts(self, feed: Dict[str, float], temp: float) -> None:
...
 
         Parameters
         ----------
-        feed: Dict[str, float]
-            instance of the Feed class
         methane_model : str
-            methane model used for methane emission calculations
+            Methane model used for methane emission calculations, including Boadi, IPCC.
+        nutrient_amount : Dict[str, float]
+            Amount of nutrients in pen ration, calculated per animal, see Notes section for units.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/calf.py]:
```suggestion
            Amounts of nutrients in pen ration, calculated per animal, see Notes section for units.
```

File: RUFAS/routines/animal/life_cycle/calf.py [hunk 15]
```diff
@@ -138,31 +138,49 @@ def calc_nutrient_rqmts(self, feed: Dict[str, float], temp: float) -> None:
...
+        nutrient_conc : Dict[str, float]
+            Concentration of nutrients in pen ration, calculated per animal, percentage.
```
Comment (ew3361zh) [inline RUFAS/routines/animal/life_cycle/calf.py]: If you think this is unnecessary, please feel free to disregard. It's just a side effect of having a partner that's an English major.
```suggestion
            Concentrations of nutrients in pen ration, calculated per animal, percentage.
```

**Comment by ew3361zh** [comment]
MyPy errors:

>In the file Pen.py, in _calc_animal_manure_excretion, it gives a complex error for the call to calc_manure_excretion, because there is a separate method for each animal class, and for lactating cows there are more input args required than for other classes. This flags a Mypy error where there are too many calls for Calf/HeiferI-III objects, but too few for cows. This logic is handled by _calc_animal_manure_excretion, where the animal object is routed to the correct call setup, but Mypy flags it anyway.
Question: Should this method be A) refactored in this PR, B) should I open an Issue, or C) this Mypy error should be flagged to be ignored?

For this one I'd probably just ignore it for now. I think it could potentially be fixed with Allister's refactoring of the animal types and how they're inheriting from one another.

>The method calc_manure_excretion in Cow.py gives the following error: Signature of "calc_manure_excretion" incompatible with supertype "HeiferIII"Mypy[override](https://mypy.readthedocs.io/en/latest/_refs.html#code-override)
Question: Should this error be officially ignored (e.g. flagging it in the code) while any changes to Animal objects (specifically related to inheritance) are made, or should we simply informally ignore it (e.g. allow the error, and open an Issue)?

I think similarly for this one. It should be ignored and we return back to it after Allister's refactoring of the animal types.

**Review by tomhuhh — APPROVED** [review]
Thanks Joe! Great work!

**Pull request merged**